lib/launcher/test/unit/Cardano/StartupSpec.hs

```diff
@@ -32,19 +33,55 @@ import Test.Hspec
 import Test.Utils.Trace
     ( captureLogging )
 import Test.Utils.Windows
-    ( nullFileName )
+    ( nullFileName, pendingOnWindows )
+
+#if defined(mingw32_HOST_OS)
+import Control.Concurrent
+    ( forkIO )
+import Control.Concurrent.MVar
+    ( MVar, newEmptyMVar, putMVar, takeMVar )
+import Control.Exception
+    ( IOException, catch )
+#endif
+
+import qualified Data.ByteString as BS
 
 spec :: Spec
 spec = describe "withShutdownHandler" $ do
     let decisecond = 100000
 
     describe "sanity tests" $ do
+        -- check file handle for input, allowing interruptions on windows
+        -- example
+        -- https://github.com/input-output-hk/ouroboros-network/blob/69d62063e59f966dc90bda5b4d0ac0a11efd3657/Win32-network/src/System/Win32/Async/Socket.hs#L46-L59
+        let hWait :: Handle -> IO Bool
+#if defined(mingw32_HOST_OS)
+            hWait h = do
+                v <- newEmptyMVar :: IO (MVar (Either IOException Bool))
+                _ <- forkIO ((hWaitForInput h (-1) >>= putMVar v . Right) `catch` (putMVar v . Left))
+                takeMVar v >>= either throwIO pure
+#else
+            hWait h = hWaitForInput h (-1)
+#endif
+        it "race hWaitForInput stdin" $ do
+            res <- race (hWait stdin) (threadDelay decisecond)
+            res `shouldBe` Right ()
+
+        it "race hWaitForInput pipe" $ withPipe $ \(a, _) -> do
+            res <- race (hWait a) (threadDelay decisecond)
+            res `shouldBe` Right ()
+
+        let getChunk :: Handle -> IO BS.ByteString
+            getChunk h = BS.hGet h 1000
+
         it "race stdin" $ do
-            res <- race (BS.hGet stdin 1000) (threadDelay decisecond)
+            pendingOnWindows "deadlocks on windows"
+            res <- race (getChunk stdin) (threadDelay decisecond)
             res `shouldBe` Right ()
 
         it "race pipe" $ withPipe $ \(a, _) -> do
-            res <- race (BS.hGet a 1000) (threadDelay decisecond)
+            pendingOnWindows "deadlocks on windows"
+            res <- race (getChunk a) (threadDelay decisecond)
```

Comment:
So we can't simply read chunks from pipes on Windows?

Reply:
Yeah actually we can, it's fine as long as it's wrapped in `forkIO`. I changed the `hWaitForInput` calls back to `hGet`.